akka-actor/src/main/scala/akka/util/MessageBuffer.scala

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright (C) 2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.util
+
+import akka.actor.ActorRef
+
+/**
+ * A single producer single consumer mutable message buffer
+ */
+final class MessageBuffer private (
+  private var _head: MessageBuffer.Node,
+  private var _tail: MessageBuffer.Node) {
+  import MessageBuffer._
+
+  private var _size: Int = if (_head eq null) 0 else 1
+
+  def isEmpty: Boolean = _head eq null
+
+  def nonEmpty: Boolean = !isEmpty
+
+  def size: Int = _size
+
+  def append(message: Any, recipient: ActorRef): MessageBuffer = {
```

Comment:
Buffers often used as a stash also, and then `recipient` is typically the `sender`. Perhaps rename to `ref`?